pkg/curatedpackages/bundleclient.go

```diff
@@ -0,0 +1,172 @@
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	packagesv1 "github.com/aws/eks-anywhere-packages/api/v1alpha1"
+	"github.com/aws/eks-anywhere-packages/pkg/artifacts"
+	"github.com/aws/eks-anywhere-packages/pkg/bundle"
+	"github.com/aws/eks-anywhere/pkg/constants"
+)
+
+// BundleClient abstracts the source of package bundle information.
+//
+// Whether the source is a registry, a cluster, the BundleClient should
+// abstract the details of retrieving a particular curated packages bundle.
+type BundleClient interface {
+	ActiveOrLatest(ctx context.Context) (*packagesv1.PackageBundle, error)
+}
+
+// BundleClientOptions alter the behavior of the bundle client when retrieving
+// package bundle information.
+type BundleClientOptions struct {
+	// ClusterName identifies the cluster with which to communicate.
+	ClusterName string
+	// KubeConfig specifies a kubeconfig file to use.
+	//
+	// Defaults to the KUBECONFIG envrionment variable, and can fall back to
+	// ~/.kube/config or similar.
+	KubeConfig string
+	// KubeVersion specifies a kubernetes version requirement for package bundles.
+	//
+	// Required if consulting a registry for package bundle information.
+	KubeVersion string
+	// Registry as an OCI Registry URL.
+	//
+	// Defaults to the EKS-A ECR. Used when consulting a registry for package
+	// bundle information.
+	Registry string
+	// RegistryClient will be used if supplied.
+	//
+	// Specifying a custom registry client is optional, but could be needed
+	// when specifying a custom registry mirror that requires authentication.
+	RegistryClient bundle.RegistryClient
+	// restConfigurator is used to inject testing constructs.
+	//
+	// It shouldn't be required for normal use.
+	restConfigurator restConfigurator
+}
+
+// NewBundleClient constructs a BundleClient based on the value of source.
+func NewBundleClient(source BundleSource, opts BundleClientOptions) (BundleClient, error) {
+	switch source {
+	case Registry:
+		rc := opts.RegistryClient
+		if rc == nil {
+			rc = bundle.NewRegistryClient(artifacts.NewRegistryPuller())
+		}
+		kubeVersion := opts.KubeVersion
+		if kubeVersion == "" {
+			return nil, fmt.Errorf("kube version is required when source is registry")
+		}
+		registry := opts.Registry
+		if registry == "" {
+			registry = "public.ecr.aws/q0f6t3x4" // TODO pull this from the EKS-A bundle
+		}
+		return NewRegistryBundleClient(rc, registry, kubeVersion), nil
+	case Cluster:
+		kubeClient, err := newKubeClient(opts.KubeConfig, opts.restConfigurator)
+		if err != nil {
+			return nil, err
+		}
+		return NewClusterBundleClient(kubeClient, opts.ClusterName), nil
+	default:
+		return nil, fmt.Errorf("unhandled package bundle source %q", source)
+	}
+}
+
+// clusterBundleClient implements BundleClient by querying an existing cluster.
+type clusterBundleClient struct {
+	kubeClient  client.Client
+	clusterName string
+}
+
+var _ BundleClient = (*clusterBundleClient)(nil)
```

Comment:
what is this for?

Reply:
https://go.dev/doc/effective_go#blank_implements

It's a common thing you'll see a lot. In English, it says "an object of type pointer to clusterBundleClient implements the BundleClient interface". This bit of code will throw an error if the compiler determines that the previous statement is no longer true.

Reply:
No, in this case we don't need it. As I said, it helps me communicate intent, and identify problems faster. I'll remove it though.